### ISSUE TITLE
Support for batches of distributions

### ISSUE THREAD

**Issue by fedario** [JuliaStats/Distributions.jl#1248]
Hey!

[PyTorch's distributions](https://pytorch.org/docs/stable/distributions.html) provides a way of generating a batch of distributions by passing the distribution parameters with an additional batch dimension.
In Distributions.jl this could look like this:

```Julia
dist = Normal(
    [1, 2, 3],
    [3, 2, 1],
)
rand(dist) 
``` 

The call to rand would return a sample of size (3,) where the  first element would come from Normal(1, 3), the second from Normal(2, 2) and the third from Normal(3, 1).
The same concept applies to multivariate distributions.
[More info on this design here](https://bochang.me/blog/posts/pytorch-distributions/).

The advantage of this design is that it makes it easy to generate a batch of distributions and samples with CUDA arrays, since these discourage scalar access.

My questions:

Is there another idiomatic way to do this with Distributions.jl?
Do you think implementing a similar design to the one used by PyTorch makes sense for Distributions.jl?

**Comment by devmotion** [comment]
You can use `map` or broadcasting to achieve the same behaviour. For instance, you can run
```julia
rand.(Normal.([1, 2, 3], [3, 2, 1]))
```
However, I have never tried this on GPUs - I would assume that in general there are problems with sampling even from single distributions on the GPU, which then of course also apply to the batches of distributions.

**Comment by fedario** [comment]
> 
> 
> You can use `map` or broadcasting to achieve the same behaviour. For instance, you can run
> 
> ```julia
> rand.(Normal.([1, 2, 3], [3, 2, 1]))
> ```
> 
> However, I have never tried this on GPUs - I would assume that in general there are problems with sampling even from single distributions on the GPU, which then of course also apply to the batches of distributions.

**Comment by fedario** [comment]
Yea, I've tried that with CUDA arrays and MvNormal, by using eachcol on the parameters and got the "scalar access warning" from CUDA, meaning that it's not performant.

**Comment by devmotion** [comment]
This is not related to batching distributions though but due to indexing operations in the implementation of `unwhiten!` in PDMats, e.g. in https://github.com/JuliaStats/PDMats.jl/blob/d50e35a37d218f0e6ac552512bc4c79b102cc930/src/pdiagmat.jl#L81. This specific example could be fixed by replacing the for-loop with a broadcast operation `@. r = x * sqrt(v)`.

**Comment by fedario** [comment]
> This is not related to batching distributions though but due to indexing operations in the implementation of unwhiten! in PDMats, e.g. in https://github.com/JuliaStats/PDMats.jl/blob/d50e35a37d218f0e6ac552512bc4c79b102cc930/src/pdiagmat.jl#L81. This specific example could be fixed by replacing the for-loop with a broadcast operation @. r = x * sqrt(v).

Wow, I'm pretty impressed you figured that out based on my mediocre problem description!

So it seems like CuArrays can in fact not be treated like regular arrays in this case, unlike their common supertype might imply. But maybe I'm overly pessimistic here.

I don't know if you are the right person to ask, but would it make sense to make PDMats compatible with CuArrays by replacing scalar accesses with broadcast ones?

**Comment by devmotion** [comment]
> So it seems like CuArrays can in fact not be treated like regular arrays in this case, unlike their common supertype might imply. But maybe I'm overly pessimistic here.

In the case of (multivariate) normal distribution it might be easier to use `CUDA.randn` directly, e.g. for diagonal covariance matrices
```julia
julia> μs = CUDA.randn(4,10)
4×10 CuArray{Float32,2}:
 0.0476941  -0.264443   0.499054   0.350746  -1.9359     0.0101001  -0.50966   -1.69315   0.377177  -0.311695
 0.985609    1.62916   -0.930584  -0.717816  -0.447085  -0.821012   -0.387925  -0.709288  0.247434  -1.31046
 1.03467    -0.518602  -0.663794   0.762055   1.82921   -0.30582     1.04381    1.50052   0.636448  -0.617996
 0.757833   -1.4049    -0.716513   2.0175     0.805598   0.659108   -0.520839   1.06928   0.822147   0.112952

julia> vars = CUDA.rand(4,10)
4×10 CuArray{Float32,2}:
 0.039748  0.111466  0.15118   0.67902   0.752794  0.701304   0.293771  0.498601   0.513519  0.544787
 0.614878  0.105713  0.850267  0.96418   0.489081  0.0756312  0.53497   0.0471762  0.905324  0.204301
 0.638271  0.367247  0.967387  0.412682  0.580954  0.832263   0.884396  0.295302   0.419665  0.104343
 0.511243  0.759502  0.46643   0.952843  0.138085  0.105386   0.88023   0.597133   0.501077  0.979644

julia> μs .+ vars .* CUDA.randn(4,10)
4×10 CuArray{Float32,2}:
 0.0559772  -0.320123    0.455332  -0.495313  -2.63995    0.237539  -0.427061  -1.96411   1.0853     -0.733353
 1.00205     1.76148    -2.0842     0.304815  -0.675685  -0.818722  -0.266044  -0.699681  0.177612   -1.42312
 1.77151    -0.0497113  -0.889992   0.570649   1.84062   -0.768166   1.32019    1.39911   0.0152018  -0.414656
 0.237956   -1.40983    -0.989351   0.52414    0.830402   0.590989  -0.159707   1.02358   0.413332   -1.62367
```

> I don't know if you are the right person to ask, but would it make sense to make PDMats compatible with CuArrays by replacing scalar accesses with broadcast ones?

Personally I think it would make sense but you would have to ask one of the maintainers of PDMats. I guess it would also be useful to check if this change fixes the example or if there are other issues that have to be resolved.

**Comment by fedario** [comment]
> In the case of (multivariate) normal distribution it might be easier to use `CUDA.randn` directly, e.g. for diagonal covariance matrices

Yea, I did something similar, which is basically the reparameterization trick, which you may have heard about, given your area of study (and the fact that you used the same technique). Still, it would be much more convenient if Distributions.jl and its dependencies worked better on the GPU, which is something I'd like to help with. 

> Personally I think it would make sense but you would have to ask one of the maintainers of PDMats. I guess it would also be useful to check if this change fixes the example or if there are other issues that have to be resolved.

Good point.

**Comment by devmotion** [comment]
Definitely, ideally Distributions and its dependencies would just work on the GPU. Unfortunately, in general quite many workarounds and fixes are needed currently. You can find some more information in the following discussion: https://github.com/cambridge-mlg/Covid19/issues/5
